File: src/main/java/org/library/Borrower.java
package org.library;

import java.util.ArrayList;
import java.util.List;

public class Borrower extends User{

    BorrowedBooks borrowedBooks;
    BookDetails curHold;

    public Borrower(String username, String password){
        super(username, password);
        borrowedBooks = new BorrowedBooks();
    }

    public BorrowedBooks getBorrowedBooks() {
        return borrowedBooks;
    }

    public List<Book> getBorrowedBooksList(){
        return borrowedBooks.getBorrowedBooksList();
    }

    public boolean hasBook(String bookTitle){
        return borrowedBooks.getBookByTitle(bookTitle) != null;
    }
    public boolean hasHold(){ return curHold != null; }
    public boolean curHoldIs(String bookTitle){
        if (!hasHold()) return false;
        return curHold.getTitle().equals(bookTitle);
    }

    public void addBorrowedBook(Book b){
        borrowedBooks.addBook(b);
    }

    public void removeBorrowedBook(String bookTitle){
        borrowedBooks.removeBook(bookTitle);
    }

    public int getBorrowedBooksNum(){ return borrowedBooks.size(); }

    public void setCurHold(BookDetails book){ curHold = book; }

    public BookDetails getCurHold(){ return curHold; }
}

File: main/Borrower.js
import { User } from "./User.js";
import BorrowedBooks from "./BorrowedBooks.js";
import BookDetails from "./BookDetails.js";

export default class Borrower extends User {
  constructor(username, password) {
    super(username, password);
    this.borrowedBooks = new BorrowedBooks();
    this.curHold = null;
  }

  getBorrowedBooks() {
    return this.borrowedBooks;
  }

  getBorrowedBooksList() {
    return this.borrowedBooks.getBorrowedBooksList();
  }

  hasBook(bookTitle) {
    return this.borrowedBooks.getBookByTitle(bookTitle) !== null;
  }

  hasHold() {
    return this.curHold != null;
  }

  curHoldIs(bookTitle) {
    if (!this.hasHold()) return false;
    return this.curHold.getTitle() === bookTitle;
  }

  addBorrowedBook(b) {
    this.borrowedBooks.addBook(b);
  }

  removeBorrowedBook(bookTitle) {
    this.borrowedBooks.removeBook(bookTitle);
  }

  getBorrowedBooksNum() {
    return this.borrowedBooks.size();
  }

  setCurHold(book) {
    this.curHold = book;
  }

  getCurHold() {
    return this.curHold;
  }
}
